### ISSUE TITLE
Implement Async for native Spark Operators

### ISSUE THREAD

**Issue by phanikumv** [astronomer/astronomer-providers#119]

**Comment by kaxil** [comment]
If we have too many issues with it, we can park it for now, or use LivyOperator as an alternative to submit Spark jobs

**Comment by phanikumv** [comment]
We've got the local connection working for the SparkSubmitOperator yesterday post further analysis. I think this is story is unblocked for now @sunank200 , @bharanidharan14 - please comment.

**Comment by sunank200** [comment]
Yes, `spark-submit` works for the deploy-mode `local`. So `SparkSubmitOperator` example dag works with local spark setup. For `SparkSqlOperator` i am setting up the hive along spark to create a table to test the example DAG.

**Comment by bharanidharan14** [comment]
on 7-03-2022 and 21-03-2022 Successfully Installed spark in the airflow worker to run the spark-submit job and Created the spark cluster container tried running the sample spark job via airflow worker, I was facing some issues with container resource allocation issue.
22-03-2022 Tried creating separate docker image, created spark master, worker cluster container and tried submitting the job via airflow worker it was able to pick up the job but there is python version mismatch

**Comment by bharanidharan14** [comment]
Working on implementing Spark submit operator async

**Comment by sunank200** [comment]
`SparkSqlOperator` requires the `spark-sql` to work when it's installed along in the docker. Right now, for deployment mode as `local`, the `SparkSqlOperator` fails as `spark-sql`  fails with the following error:

> `22/03/24 05:12:55 ERROR Utils: Uncaught exception in thread shutdown-hook-0
> java.lang.InterruptedException
>         at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedNanos(AbstractQueuedSynchronizer.java:1039)
>         at java.util.concurrent.locks.AbstractQueuedSynchronizer.tryAcquireSharedNanos(AbstractQueuedSynchronizer.java:1328)
>         at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:277)
>         at org.sparkproject.jetty.server.AbstractConnector.doStop(AbstractConnector.java:371)
>         at org.sparkproject.jetty.server.AbstractNetworkConnector.doStop(AbstractNetworkConnector.java:88)
>         at org.sparkproject.jetty.server.ServerConnector.doStop(ServerConnector.java:246)
>         at org.sparkproject.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:94)
>         at org.sparkproject.jetty.server.Server.doStop(Server.java:459)
>         at org.sparkproject.jetty.util.component.AbstractLifeCycle.stop(AbstractLifeCycle.java:94)
>         at org.apache.spark.ui.ServerInfo.stop(JettyUtils.scala:525)
>         at org.apache.spark.ui.WebUI.$anonfun$stop$2(WebUI.scala:174)
>         at org.apache.spark.ui.WebUI.$anonfun$stop$2$adapted(WebUI.scala:174)
>         at scala.Option.foreach(Option.scala:407)
>         at org.apache.spark.ui.WebUI.stop(WebUI.scala:174)
>         at org.apache.spark.ui.SparkUI.stop(SparkUI.scala:101)
>         at org.apache.spark.SparkContext.$anonfun$stop$6(SparkContext.scala:2071)
>         at org.apache.spark.SparkContext.$anonfun$stop$6$adapted(SparkContext.scala:2071)
>         at scala.Option.foreach(Option.scala:407)
>         at org.apache.spark.SparkContext.$anonfun$stop$5(SparkContext.scala:2071)
>         at org.apache.spark.util.Utils$.tryLogNonFatalError(Utils.scala:1442)
>         at org.apache.spark.SparkContext.stop(SparkContext.scala:2071)
>         at org.apache.spark.sql.hive.thriftserver.SparkSQLEnv$.stop(SparkSQLEnv.scala:77)
>         at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver$.$anonfun$main$2(SparkSQLCLIDriver.scala:143)
>         at org.apache.spark.util.SparkShutdownHook.run(ShutdownHookManager.scala:214)
>         at org.apache.spark.util.SparkShutdownHookManager.$anonfun$runAll$2(ShutdownHookManager.scala:188)
>         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
>         at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:2019)
>         at org.apache.spark.util.SparkShutdownHookManager.$anonfun$runAll$1(ShutdownHookManager.scala:188)
>         at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
>         at scala.util.Try$.apply(Try.scala:213)
>         at org.apache.spark.util.SparkShutdownHookManager.runAll(ShutdownHookManager.scala:188)
>         at org.apache.spark.util.SparkShutdownHookManager$$anon$2.run(ShutdownHookManager.scala:178)
>         at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
>         at java.util.concurrent.FutureTask.run(FutureTask.java:266)
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
>         at java.lang.Thread.run(Thread.java:748)`

Trying to debug the issue right now and blocked on this.

**Comment by phanikumv** [comment]
@sunank200  lets stop further effort on SparkSQLOperator and you can start on LivyOperator

**Comment by bharanidharan14** [comment]
- Added Spark submit async Operator, Trigger, Hooks
- Added Test case, Mypy, doc string

**Comment by bharanidharan14** [comment]
Added the changes to submit the spark submit job in execute method and getting the status from trigger

**Comment by phanikumv** [comment]
Below operators are on hold due to connectivity issue between airflow worker container and EMR Spark
- SparkSubmitOperator
- SparkSqlOperator
- SparkJDBCOperator
- SparkKubernetesOperator
